examples/systemcalls/systemcalls.c

```diff
@@ -92,6 +152,68 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
  *   The rest of the behaviour is same as do_exec()
  *
 */
+    pid_t child_pid;
+
+    int fd = open( outputfile, O_WRONLY | O_TRUNC | O_CREAT, 0644 );
+
+    if( fd == -1 )
+    {
+        perror("open_file");
+        return false;
+    }
+
+    child_pid = fork();
+
+    if( child_pid == -1 )
+    {
+        perror("fork");
+        return false;
+    }
+    
+    if( child_pid == 0 )
+    {
+        if (dup2(fd, 1) < 0) 
+        { 
+            perror("dup2");
+            close(fd);
+            return false;
+        }
+        close(fd);
+        execv(command[0], command); 
+        perror("execv");
+        exit(1);
+    }
+    
+    int status;
+    pid_t wait_ret = waitpid( child_pid, &status, 0 );
+
+    if( wait_ret == -1 )
+    {
+        perror("wait");
+        return false; 
+    }
+    else
+    {
+        // check if child process exited
+        if( WIFEXITED(status))
+        {
+            // check for exit status 0 for success
+            if( WEXITSTATUS(status) == 0)
+            {
+                return true;
+            }
+            else
+            {
+                return false;
```

Comment:
Missing cleanup in these code paths.
Also see discussion in class about avoiding duplication between these two very similar functions.

Reply:
@dwalkes Yes I realized it after the discussion in class. I will keep in mind for the future assignments.